### DISCUSSION TITLE
NextJS 13 Custom JWT decode method middleware causes error: "oidc-token-hash Cannot read properties of undefined"

### DISCUSSION THREAD

**Discussion by 6215x** [nextauthjs/next-auth#6297] [Help]
**The Problem:**
NextJS 13.1.1 is what I'm using. When using Custom JWT decode method and pass the same decode method to withAuth in Middleware.ts cause error: "oidc-token-hash Cannot read properties of undefined"

<img width="605" alt="image" src="https://user-images.githubusercontent.com/60601102/210791434-042845c8-ee2f-4036-9d15-e88e562ff82d.png">

My **middleware.ts**
```
import { withAuth } from "next-auth/middleware"
import { authOptions } from "./pages/api/auth/[...nextauth]"

export const config = {
    matcher: [
        "/",
        "/dashboard",
    ]
}
export default withAuth({
    jwt: {
        decode: authOptions.jwt as any,
    },
    callbacks: {
        authorized: ({ req, token }) => !!token,
    },
    pages: {
        signIn: '/auth/login',
    },
})
```

**Comment by seanparkross** [comment]
Exact same problem. My decode function uses the [jose](https://www.npmjs.com/package/jose) package but don't think that's the issue.

This is my decode function: 

```typescript
decode: async ({ secret, token }: JWTDecodeParams) => {
  const { payload } = await jwtVerify(token as string, new TextEncoder().encode(secret as string));
  return (payload as JWT) ?? null;
},
```

**Reply by seanparkross** [reply]
@Gruszht6215 Did you figure anything out with this?

**Reply by 6215x** [reply]
No, I still can't figure it out

**Reply by firmanjml** [reply]
I solved this by creating a new constant outside [...nextauth].ts ie /utils/authOption.ts 
containing 

```
const cookies = {
    sessionToken: {
        name: "appName.session-token",
        options: {
            httpOnly: true,
            secure: false,
            path: "/"
        }
    },
    csrfToken: {
        name: "appName.csrf-token",
        options: {
            httpOnly: true,
            secure: false,
            path: "/"
        }
    },
    callbackUrl: {
        name: "appName.callback-url",
        options: {
            httpOnly: true,
            secure: false,
            path: "/"
        }
    },
}

const pages = {
    signIn: "/"
}

export const AUTH_OPTIONS = {
    pages,
    cookies
}
```

then import it on [...nextauth].ts and middleware 

middleware.ts
```
export default withAuth({
    cookies: AUTH_OPTIONS.cookies,
    pages: AUTH_OPTIONS.pages
});
```

[...nextauth].ts
```
cookies: AUTH_OPTIONS.cookies,
pages: AUTH_OPTIONS.pages,
```

I found that importing the propeties  directly from [...nextauth] in middleware can cause this issue

**Reply by thexpand** [reply]
I tried this but it doesn't seem to work with the new Next 13 route handlers API in `/app/api`. How did you resolve it?
